### ISSUE TITLE
Set protocol (and perhaps others) errors are silently ignored

### ISSUE THREAD

**Issue by bungle** [ada-url/ada#719]
### Version

2.9.1

### Platform

Darwin kong-222.local 23.6.0 Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:30 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6000 arm64 arm Darwin

### What steps will reproduce the bug?

```
ada_set_href(?, "unknown:", 8) -- returns true
ada_set_protocol(?, "https:", 6) -- returns true
ada_get_protocol(?) -- -- returns "unknown:" 
ada_get_href(?) -- returns "unknown:"
```

### How often does it reproduce? Is there a required condition?

Always.

### What is the expected behavior?

This is a bit more complicated. I think especially the `setters` behave a bit undetermistic (or hard to follow) depending on what was set before or when the url was parsed. I feel that the state is not correctly updated. Perhaps the validations run in wrong place, aka in `setters` (e.g. if you were to change `file:///blaa` to `https://user:pass@example.com/` what is the sequence of setters you need to call, pseudo:
```
u = ada_parse("file:///blaa")
set_protocol(u, "https:")
set_username(u, "user")
set_password(u, "pass")
set_hostname(u, "example.com")
set_pathname(u, "/")
```

Or vise-versa? You cannot set username or password to `file:`, but you can to `https:`. If validating during the `sets` then you need to know which order is right, and which parameters are required, e.g. is `protocol` a pre-requisite to validate `set_username`, and thus `set_protocol` should be called before `set_username` but then again setting protocol to `file:` may require clearing username and password before setting protocol. But as following shows, there is more to it, as `set_protocol` seems to return `true` but not set the protocol.

I am asking this as I try to compose urls from parameters like:
```
{
  protocol = "https:",
  username = "user",
  password = "pass",
  host = "example.com:1234",
  hostname = "example.com",
  port = 1234,
  pathname = "/foo/bar",
  search = "?baz",
  hash = "#quux",
}
```

Perhaps additional `parse_structured` could help on this, but it does not solve above issues fully.

### What do you see instead?

Please see the "discussion" in `expected behavior`.

**Comment by anonrig** [comment]
I think you have some assumptions that does not conform to the WHATWG specification. Let me explain, through for example code:

> u = ada_parse("file:///blaa")

Now the valid url is `file:///blaa`

> set_protocol(u, "https:")

This would fail because `https:///blaa` is not a valid URL.

> set_username(u, "user")

Remember our URL is still `file:///blaa`. File URL's can not have usernames, since file URL's does not have hostname, but pathname. For example `file://username@/blaa` is not a valid URL

> set_password(u, "pass")

Similar to set_username. File URLs can not have passwords.

> set_hostname(u, "example.com")

Setting hostname would make the URL: `file://example.com/blaa`

> set_pathname(u, "/")

Setting pathname would make the result `file://example.com/`

**Comment by bungle** [comment]
> This would fail because `https:///blaa` is not a valid URL.

But it does not as explained above. The `set_protocol` returns `true` like it succeed.

**Comment by anonrig** [comment]
Ah I see. This is clearly a bug on the `set_protocol()` function. The specification does not mention anything about setters returning boolean, that's our personal touch, and clearly there is a lack of test coverage there. Thanks for reporting!

**Comment by bungle** [comment]
> Ah I see. This is clearly a bug on the `set_protocol()` function. The specification does not mention anything about setters returning boolean, that's our personal touch, and clearly there is a lack of test coverage there. Thanks for reporting!

I didn't test further, but I have a feeling that this may not be the only one. I was even thinking that does the validation make sense at all on setters. Perhaps just set and then people should call is_valid to check if url is really valid.

**Comment by anonrig** [comment]
fyi: setters would never make an URL `invalid` by setting it as `is_valid = false` that is against the specification. I think we are just not returning the correct succeed value correctly.

**Comment by bungle** [comment]
> fyi: setters would never make an URL `invalid` by setting it as `is_valid = false` that is against the specification. I think we are just not returning the correct succeed value correctly.

Yes, but then you endup with this problem that how to set structured input. You cannot create empty `url` and then call setters. You need to create something.

**Comment by anonrig** [comment]
Yes, this was also mentioned in https://github.com/ada-url/ada/issues/573. URL specification does not allow us to construct URLs from scratch.

**Comment by bungle** [comment]
@anonrig didn't check but does the #573 mean that this does not work:
```
ada::url url;
url.set_protocol("https");
url.set_host("google.com");
return url.get_href();
```

but this may work?
```
ada::url url;
url.set_host("google.com");
url.set_protocol("https");
return url.get_href();
```

**Comment by bungle** [comment]
I tried this with my Lua bindings:

```lua
local ada = require "resty.ada"

local url = ada.parse("file:")
print(url:set_host("google.com"))
print(url:set_protocol("https"))
print(url:get_href()) -- outputs: https://google.com/

local url = ada.parse("file:")                                                                                          
print(url:set_protocol("https"))
print(url:set_host("google.com"))                                                                                             
print(url:get_href()) -- outputs: file://google.com/
```

All `sets` return `true` but results are:

**Comment by lemire** [comment]
@bungle Thanks for the detailed reports.
